Use Ruby implementation for `which`

davispuh opened the pull request:
In Windows there's no `/dev/null` and it causes error "The System cannot find the path specified". Also in Windows there's no native `which` and it works only because it's included in DevKit.

This Ruby implementation should correctly work for all platforms.

djberg96 commented:
Please merge.

tduehr commented:
Why iterate over the path and extensions rather than use a glob?

djberg96 commented:
@tduehr wouldn't work on Windows with their file extensions, e.g. notepad.exe, not notepad.

tduehr commented:
`Dir.glob("*.{exe,bat}")` works for me in 2.2.2 (compiled w/ VC) under win8.1. `C:\Ruby\bin`

`Dir.glob("**/*{#{ENV["PATHEXT"].tr(';',',')}}")` works in the directory above it `C:\Ruby`

djberg96 commented:
Hm, for some reason I thought that approach had some issue, but perhaps I'm mistaken. Anyway, it's not like we need to hyper optimize for this little bit of functionality. I'm happy with whatever works. :)

davispuh commented:
I'm don't really understand why glob would be preferable? I'm not sure if there would be any performance benefits.
Here's implementation using Dir.glob and it seems to be more complicated...
```ruby
def which(name)
  exts = ENV['PATHEXT'] ? '{' + ENV['PATHEXT'].split(';').join(',') + '}' : ''
  paths = '{' + ENV['PATH'].split(File::PATH_SEPARATOR).join(',').gsub('\\', '/') + '}'
  Dir.glob(paths + '/' + name + exts).each do |app|
    return app if File.executable?(app)
  end
  nil
end
```

In PATH there can be multiple executables with same name and some might not even be executable so we've to check all in correct order.

tduehr commented:
I was wondering if there was a specific reason is all... Then i got to way over optimizing. I'll pull this when I get a chance to write something to use which normally then failover to this.
